### PR TITLE
feat(authN): Redesign JWT token auth

### PR DESCRIPTION
**Pull request by michalkrzyz**
Redesign JWT token authentication middleware to support additional/alternative authentication method

## Description
Implementation of the capability of having multiple auth mechanisms in place.
To register new auth method (JWT or oidc), function with Verify(c *gin.Context) error has to be implemented.
Currently only JWT token based authentication is used, oidc implementation with it's Verify method will be implemented in separate issue/PR.

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [x] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

- Related Issue # (issue)
- Closes # (issue)
- Fixes # (issue)

> Remove if not applicable

## Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help
- [ ] Separate ticket for tests # (issue/pr)

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Added to documentation?

- [x] 📜 README.md
- [ ] 🤝 Documentation pages updated
- [ ] 🙅 no documentation needed
- [ ] (if applicable) generated OpenAPI docs for CRD changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: internal/api/graphql/access/token_auth_method.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+type TokenClaims struct {
+	Version string `json:"version"`
+	jwt.RegisteredClaims
+}
+
+type TokenAuthMethod struct {
+	logger Logger
+	secret []byte
+}
+
```
Comment (drochow) [inline internal/api/graphql/access/token_auth_method.go:41]: A  Docstring would probably be good to say: 
- Verifies token presence in header
- Validates token by parsing
- Checks token validity (signature)
- Verifies expiration time

Due to implementation constraints, we intentionally do: 
- No validation of the token issuer (iss claim)
- No validation of the "not before" time (nbf claim)
- No audience validation (aud claim)
- No token signing algorithm verification before parsing

**Review by drochow — CHANGES_REQUESTED** [review]
Good Job!

LGTM --> i would just like to have some specific comments to the verify function that make clear that we left out certain enforcements/verifications intentionally

File: internal/api/graphql/access/token_auth_method.go (hunk 1 above)
Comment (michalkrzyz) [reply internal/api/graphql/access/token_auth_method.go:41]: I am not a big fan of commenting the code ("good code documents itself" :) ). Would it be ok if I enclose each check in single responsibility functions like:
checkHeaderInRequest
verifyTokenStructure
checkTokenSignature
verifyTokenExpiration
etc.

Comment (michalkrzyz) [reply internal/api/graphql/access/token_auth_method.go:41]: @drochow please check the last commit. I have added some functions to make checks/gets more readable.

**Review by MR2011 — APPROVED** [review]

File: internal/api/graphql/access/test/util.go [hunk 2]
```diff
@@ -81,7 +81,7 @@ func (j *Jwt) String() string {
 			ExpiresAt: j.expiresAt,
 			IssuedAt:  jwt.NewNumericDate(time.Now()),
 			Issuer:    "heureka",
```
Comment (dorneanu) [inline internal/api/graphql/access/test/util.go:83]: This should also be configurable. Most probably from ENV variables.

File: internal/api/graphql/access/auth.go [hunk 3]
```diff
@@ -17,23 +19,52 @@ type Logger interface {
...
+	auth := Auth{logger: l}
+	auth.AppendInstance(NewTokenAuthMethod(l, cfg))
+	//TODO: auth.AppendInstance(NewOidcAuthMethod(l, cfg))
+	return &auth
 }
 
-func NewAuth(cfg *util.Config) Auth {
-	l := newLogger()
+type Auth struct {
+	chain  []AuthMethod
```
Comment (dorneanu) [inline internal/api/graphql/access/auth.go:31]: Maybe I'm missing the full picture, but why exactly do we need a chain of auth methods?

**Comment by dorneanu** [comment]
In the description I cannot see the "related issues" nor the ones which the PR should close.

File: internal/api/graphql/access/auth.go (hunk 3 above)
Comment (dorneanu) [reply internal/api/graphql/access/auth.go:31]: Hmmm.. Even if OIDC uses JWTs as part of its specification, we do need to treat both auth mechanisms differently. I mean, you also have different fields (inside the JWT token payload, see this [example](https://connect2id.com/learn/openid-connect#:~:text=The%20ID%20token%20statements%2C%20or%20claims%2C%20are%20packaged%20in%20a%20simple%20JSON%20object%3A)). Or how is this supposed to work? Maybe we need some call to clarify this.

**Review by dorneanu — APPROVED** [review]
Muy bien

**Review by drochow — APPROVED** [review]

**Pull request merged**